### ISSUE TITLE
[Enhancement]

### ISSUE THREAD

**Issue by kduntz** [aws-cloudformation/cloudformation-guard#312]
**Is your feature request related to a problem? Please describe.**

Many CloudFormation resource schemas use String for values that would probably be better represented as Integers. An example is "MinSize" from [AWS::AutoScaling::AutoScalingGroup](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-group.html#cfn-as-group-minsize). This does not work out well when trying to write rules that use binary operators for comparison such as ">=". For example, if we wanted to enforce a "MinSize" greater than three and a customer configured "12", we have problems as String "3" will be considered greater than String "12".

**Describe the solution you'd like**

A way to compare strings in a natural way. It seems like being able to parse int from string would be enough to unblock our use cases, though I expect this is somewhat short sighted.

**Describe alternatives you've considered**

I took a peek at the examples and sources available and didn't find any working alternatives. Our only option at the moment is to either forego writing rules on these sort of values or come up with our own solution outside of Guard.

**Comment by joshfried-aws** [comment]
Hi @kduntz thanks for reaching out. Now that we have introduced support for some builtin functions I think we could leverage those to accomplish a solid developer experience here. 

What would you think about having functions that do conversions (similar to how they're done in languages like python). This would enable us to convert string values to numbers/booleans, and numbers/booleans to strings, etc. The developer could then take the result of the operation, and call whichever comparison operator they'd like on it.

**Comment by kduntz** [comment]
I think that sounds reasonable, something like a "parseInt" for each of the primitives would be enough since we know what they need to be converted to for proper comparison.

**Comment by joshfried-aws** [comment]
Hi @kduntz we have added functionality to convert certain types in #388. 

These are the conversions that we now support 

1. strings/floats-> ints
2. strings/ints -> floats
3. strings -> bools
4. bools/floats/ints -> strings

I will provide an example below on how to write a rule that converts a string into an int, so that we can write integer comparators with them.

Given the following template 

```
Resources:
  asg:
    Type: AWS::AutoScaling::AutoScalingGroup
    Properties:
      MaxSize: "5.0"
      MinSize: "1"
      NewInstancesProtectedFromScaleIn: true
      DefaultInstanceWarmup: 1.5
      HealthCheckGracePeriod: 1
      HealthCheckType: "true"
 ```
 
We can now use the following rule to convert the MinSize into an int, and then do integer comparisons on it.
 ```
 let asg = Resources.*[ Type == 'AWS::AutoScaling::AutoScalingGroup' ]

rule test_parse_int when %asg !empty {
    let min = parse_int(%asg.Properties.MinSize)

    %min == 1
}
```

I am going to go ahead and close this issue. Feel free to reopen it at anytime, or to create a new one for clarifications.

Thanks,

**Comment by GITOffMuhLawn** [comment]
What version are these in?  The documentation says 3.0.0 is the latest, I'm currently running that, and gives an error saying that the functions don't exist whenever I try to use them.

**Comment by razcloud** [comment]
Hi @GITOffMuhLawn,

This should be available in 3.x.  Can you please provide the exact error you are receiving as well as the template/rule so that we can re-create and triage. 

Thank you

**Comment by GITOffMuhLawn** [comment]
I upgraded to 3.0.3 and it worked.  I literally just upgraded from choco, restarted vscode for good measure, and reran the cfn-guard validate command and it worked like I was hoping it was going to.  I didn't screen shot it, but the error was "the function doesn't exist" or something very close to that language.  I didn't know 3.0.3 was the latest version because the user guide note 3.0.0.  I don't even see the parse functions documented anywhere and if it wasn't for stumbling on this randomly through a google search I would have no idea it supported it.  
The other functions (non parse functions) aren't even in the user guide.  I could understand if github was intended to be where you wanted everyone to look at everything, but all the links https://github.com/aws-cloudformation/cloudformation-guard/blob/main/docs/FUNCTIONS.md send you to the user guide, and then the link for functions send you the the .md in github so there is no consistency on what anyone is supposed to use. So you end up getting send between User Guide and Git Hub, and it looks like both are missing coverage on the documentation and the User Guide doesn't even have the current version.   

Another things that's missing in the documentation is the regex support.  I submitted something on here and got a response which was appreciated (I had figured it out prior to that).  When you use regex for querying/filtering or for a an evaluation use use / Regex here don't have to start with ^ or end with $/, but then in regex_replace you use regex like it used everywhere else.  Without anything regarding the regex support written down that's insane.  

I really think the documentation on this might be preventing larger adoption.  I've had to almost write my own manual through trial and error and combing through the examples in the user guide to understand what's going on with it.  Very basic stuff is straightforward, but as soon as you want to try and get a little creative it's extremely difficult because the documentation is all over the place (like the functions not being in the user guide from what I see).  Had I not tried to deal with regex before I used regex_replace I'd be operating under the assumption of using standard regex because that's what it looks like in the one example showing it's use.
